0006-ffmul-revised.md

```diff
@@ -0,0 +1,518 @@
+
+$$\tag{C6: high limb}
+p_2 - r_2 + p_{11} + c_0 = 2^{\ell} c_1\mod n
+$$
+
+Again, we introduce a carry value $c_1$ which is supposed to be positive, since $r_2$ has at most $\ell$ bits. Estimating the LHS shows us how many bits $c_1$ will at most have:
+
+$$
+p_2 - r_2 + p_{11} + c_0 < 6 \cdot 2^{2\ell} - 0 + 2^{\ell + 2} + 4 < 2^{2\ell + 3}
+$$
+
+We see that $c_1$ has at most $\ell + 3$ bits. To save rows and not fill more space among the permutable witness cells, we won't pass $c_1$ to a RC gate. Instead, we inline the entire RC into the ffmul gate itself, using a combination of 7x 12-bit lookups plus 3x 2-bit constraints plus 1x 1-bit constraint, giving us $7 \cdot 12 + 3 \cdot 2 + 1 = 91 = \ell + 3$ bits. We name the 11 new carry witnesses after the lowest bit of $c_1$ they cover (counting bits from 0 to 90):
+
+* **Introduce 11 witnesses**: $c_{1,0}$, $c_{1,12}$, $c_{1,24}$, $c_{1,36}$, $c_{1,48}$, $c_{1,60}$, $c_{1,72}$, $c_{1,84}$, $c_{1,86}$, $c_{1,88}$, $c_{1,90}$
+
+$$
+\begin{align}
+\tag{12-bit lookups}
+& c_{1,0}, c_{1,12}, c_{1,24}, c_{1,36}, c_{1,48}, c_{1,60}, c_{1,72} \in [0,2^{12}) \\
+\tag{C7: 2-bit}
+& c_{1,84}(c_{1,84}-1)(c_{1,84}-2)(c_{1,84}-3) = 0 \bmod n \quad \\
+\tag{C8: 2-bit}
+& c_{1,86}(c_{1,86}-1)(c_{1,86}-2)(c_{1,86}-3) = 0 \bmod n \quad \\
+\tag{C9: 2-bit}
+& c_{1,88}(c_{1,88}-1)(c_{1,88}-2)(c_{1,88}-3) = 0 \bmod n \quad \\
+\tag{C10: 1-bit}
+& c_{1,90}(c_{1,90}-1) = 0 \mod n \\
+\end{align}
+$$
+
+In place of $c_1$, the gate uses the expression
+
+$$
+c_1 = \sum_i c_{1,i} 2^{i}
+$$
+
+so $c_1$ doesn't have to be stored as a witness.
+
+By design, $c_1 \in [0, 2^{\ell + 3})$. In combination with our estimate on $p_2$ and RCs on $r_2$ and $p_{11} = p_{110} + 2^\ell p_{111}$, we can prove that $\text{(C6: high limb)}$ does not overflow $n$ and holds over the integers:
+
+$$
+|p_2 - r_2 + p_{11} + c_0 - 2^{\ell} c_1| < 6\cdot 2^{2\ell} + 2^{\ell} + 2^{\ell + 2} + 4 + 2^{2\ell + 3} < 2^{2\ell + 4} < n
+$$
+
+Therefore, we can plug in $\text{(C6)}$ into $(2'')$ to prove that
+
+$$\tag{$2'''$}
+a b - q f - r = 2^{3\ell} (w + c_1)
+$$
+
+
+
+By our first constraint $\text{(C1)}$, the LHS of $(2''')$ is a multiple of $n$, so we have
+
+
+$$
+2^{3\ell} (w + v_{1})  = 0 \mod{n}
+$$
+
+Because $2^{3\ell}$ and $n$ are co-prime, we can multiply with $2^{-3\ell}$ to see that
+
+$$
+w + v_{1}  = 0 \mod{n}
+$$
+
+Therefore, we can write $w + v_{1} = \delta n$ for some $\delta \in \mathbb{Z}$. From $(2''')$ we conclude that
+
+$$\tag{$2''''$}
+a b - q f - r = 2^{3\ell} n \delta
+$$
+
+#### Final bound
+
+There is one step remaining to make our gate sound: We must prove that $a b - q f - r$ cannot overflow $2^{3\ell} n$, so $\delta$ is in fact zero.
+
+To do this, our strategy is to add an additional bound on the highest limbs of $a$, $b$, $q$. Skipping the lower limbs and $r$ is fine because they all contribute a negligible amount to the overall estimate.
```

Comment:
Sure I can add that, in fact I think $2^{259} - 1$ is the precise maximum because you can see in one of the inequalities that the modulus needs to satisfy $2^{2\ell} (f_2 + 1) \le n$, and $2^{2\ell} \lfloor \frac{n}{2^{2\ell}} \rfloor = 2^{259}$